jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,210 @@
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+  }
+
+  /**
+   * Begins checking for an update in a separate thread.
+   *
+   * @param log {@link Consumer} used to log messages
+   * @param versionUrl the location to check for the latest version
+   * @return a new {@link UpdateChecker}
+   */
+  public static UpdateChecker checkForUpdate(Consumer<LogEvent> log, String versionUrl) {
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    Future<Optional<String>> messageFuture =
+        executorService.submit(
+            () ->
+                performUpdateCheck(
+                    log, Verify.verifyNotNull(ProjectInfo.VERSION), versionUrl, getConfigDir()));
+    executorService.shutdown();
+    return new UpdateChecker(messageFuture);
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      Consumer<LogEvent> log, String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    Path configFile = configDir.resolve("config.json");
+    Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+
+    try {
+      // Check global config
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        try {
+          ConfigJsonTemplate config =
+              JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+          if (config.disableUpdateCheck) {
+            return Optional.empty();
+          }
+        } catch (IOException ex) {
+          log.accept(
+              LogEvent.warn(
+                  "Failed to read global Jib config: "
+                      + ex.getMessage()
+                      + "; you may need to fix or delete "
+                      + configFile
+                      + "; "));
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        try (OutputStream outputStream = Files.newOutputStream(configFile)) {
+          JsonTemplateMapper.writeTo(config, outputStream);
+        } catch (IOException ex) {
+          // If attempt to generate new config file failed, delete so we can try again next time
+          log.accept(LogEvent.debug("Failed to generate global Jib config; " + ex.getMessage()));
+          Files.deleteIfExists(configFile);
+        }
+      }
+
+      // Check time of last update check
+      if (Files.exists(lastUpdateCheck)) {
+        try {
+          String fileContents =
+              new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
```

Comment:
It's easier for us to make the decision to shut it down this way.

Reply:
I just meant pass in the `Instant` as an argument?  @TadCordle had been asking about testing.  Or alternatively pass in the `Clock`.

Reply:
I commented on the wrong comment... Oops. I meant for the executor

Reply:
I commented on the wrong comment... Oops. I meant for the executor

Reply:
@briandealwis I think it might help for unit testing, but I think that's in good enough shape right now. The harder problem is integration testing.

Reply:
This class is hard to test because all of the logic is hard-coded into this single method.  Extract the last-check time determination, the current-time generation (Supplier), the modification time comparison (BiPredicate), and this class should be testable through unit tests.